### PR TITLE
feat: support OpenTelemetry metrics

### LINKED ISSUE #105: Export metrics
### Checklist

- [X] I agree to the terms within the [OpenFGA Code of Conduct](https://github.com/openfga/.github/blob/main/CODE_OF_CONDUCT.md).

### Describe the problem you'd like to have solved

As a consumer of the SDK, I would like to hook it to my dashboards to get data on several metrics.

### Describe the ideal solution

Integrate the OpenTelemetry Go API library https://github.com/open-telemetry/opentelemetry-go

Expose only from the inner OpenFgaApi. We need to be able to get these metrics:
- The latency of the request, split into
  - time it took since dev called the method until they get a response
  - time it took since the sdk issued a request to the API until it got a response
  - time reported by the server (query_duration_ms header)
- Response codes
- Error codes and method
- Method calls 

See here for examples on usage:
- datastore query count in OpenFGA https://github.com/openfga/openfga/blob/4b47d50594b9c7e436d64cf783c0aedba30b940f/pkg/server/list_users.go#L95-L98
- other examples: https://github.com/open-telemetry/opentelemetry-go/tree/main/example/otel-collector

### PR DESCRIPTION
**Pull request by evansims**
## Description

Introduces [OpenTelemetry](https://opentelemetry.io/) metrics reporting into the SDK for specific actions. We're intentionally starting with only a few metric events that we can then grow over time.

We don't have any formal documentation for this currently. We'll write this as we roll it out across all SDKs, so here is a summary of the events and the associated attributes:

### Supported Metrics

| Metric Name                     | Type      | Description                                                                          |
|---------------------------------|-----------|--------------------------------------------------------------------------------------|
| `fga-client.request.duration`   | Histogram | The total request time for FGA requests                                              |
| `fga-client.query.duration`     | Histogram | The amount of time the FGA server took to internally process nd evaluate the request |
|` fga-client.credentials.request`| Counter   | The total number of times a new token was requested when using ClientCredentials     |

### Supported attributes

| Attribute Name                 | Type     | Description                                                                                                             |
|--------------------------------|----------|-------------------------------------------------------------------------------------------------------------------------|
| `fga-client.response.model_id` | `string` | The authorization model ID that the FGA server used                                                                     |
| `fga-client.request.method`    | `string` | The FGA method/action that was performed (e.g. `check`, `listObjects`, ...) in camelCase                                |
| `fga-client.request.store_id`  | `string` | The store ID that was sent as part of the request                                                                       |
| `fga-client.request.model_id`  | `string` | The authorization model ID that was sent as part of the request, if any                                                 |
| `fga-client.request.client_id` | `string` | The client ID associated with the request, if any                                                                       |
| `fga-client.user`              | `string` | The user that is associated with the action of the request for check and list objects                                   |
| `fga-client.request.retries`   | `int`    | The number of retries attempted (starting from 1 for the original request). Deprecated, use `http.request.resend_count` |
| `http.request.resend_count`    | `int`    | The number of retries attempted (starting from 1 for the original request)                                              |
| `http.status_code`             | `int`    | The status code of the response. Deprecated, use `http.response.status_code`                                            |
| `http.response.status_code`    | `int`    | The status code of the response                                                                                         |
| `http.method`                  | `string` | The HTTP method for the request. Deprecated, use `http.request.method`                                                  |
| `http.request.method`          | `string` | The HTTP method for the request                                                                                         |
| `http.host`                    | `string` | Host identifier of the origin the request was sent to                                                                   |
| `url.scheme`                   | `string` | HTTP Scheme of the request (`http`/`https`)                                                                             |
| `url.full`                     | `string` | Full URL of the request                                                                                                 |
| `user_agent.original`          | `string` | User Agent used in the query                                                                                            |
| `http.client.request.duration` | `int`    | The total request time for FGA requests                                                                                 |
| `http.server.request.duration` | `int`    | The amount of time the FGA server took to internally process nd evaluate the request                                    |

## References

Closes #105 

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: api_open_fga.go [hunk 1]
```diff
@@ -918,6 +921,16 @@ func (a *OpenFgaApiService) CheckExecute(r ApiCheckRequest) (CheckResponse, *_ne
 			return localVarReturnValue, nil, err
 		}
 
+		var otel = &telemetry.Telemetry{}
+		var attrs = otel.Metrics().AttributesFromRequest(req)
+		attrs[telemetry.FGAClientRequestMethod] = "Check"
+		attrs[telemetry.FGAClientRequestStoreID] = r.storeId
+		attrs[telemetry.HTTPRequestResendCount] = strconv.Itoa(i + 1)
```
Comment (rhamzeh) [inline api_open_fga.go]: Can we make this start from 0, and not send it if the request has not been retried?

https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-client-retries-examples

Also here
Screenshot 2024-07-30 at 10 50 05 AM

**Review by rhamzeh — COMMENTED** [review]
This does not yet implement fga_client.user right?

**Comment by evansims** [comment]
> This does not yet implement fga_client.user right?

Correct, I haven't quite figured out how to approach that yet; I'll definitely dig into that with a separate PR, after we get the essentials here

File: api_open_fga.go (hunk 1 above)
Comment (evansims) [reply api_open_fga.go]: @rhamzeh Sure thing, easy change! I've updated the PR

**Review by JAORMX — APPROVED** [review]

**Review by ewanharris — APPROVED** [review]

**Review by rhamzeh — APPROVED** [review]

**Pull request merged**